### ISSUE TITLE
RFE: Support for index ranges and index terms in sections

### ISSUE THREAD

**Issue by apnadkarni** [asciidoctor/asciidoctor#2048]
Enhancement request - support for page ranges in an index so an entry might look like

    tables
        formatting, 145-160

One possibility for syntax is just another attribute as has been suggested for #2047.

    indexterm:[tables, formatting, range="start"]
    Lots of text over multiple pages.
    indexterm:[tables, formatting, range="end"]

What I would prefer (in lieu of or in addition to the above) because it is more convenient would be to allow index terms with the current syntax to be used in the section headers and have them automatically cover the range of pages taken up by that section.

**Comment by mojavelinux** [comment]
>  allow index terms with the current syntax to be used in the section headers and have them automatically cover the range of pages taken up by that section.

That's an interesting idea. We could add indexterms as an attribute on a section. Each term triple would be separated by semi-colons.

```
[indexterms="JPQL, query language"]
== JPQL

content
```

**Comment by mojavelinux** [comment]
For defining custom ranges using the shorthand syntax, here's my idea (something like a heredoc string):

```
content

(((Flash flash>>)))Flash has been supplanted by HTML5.

more content

more content

more content

more content(((<<flash)))
```

other ideas include:

* `(((term =begin#id))) ... (((=end#id)))`
* `(((term id=))) ... (((=id)))`
* `(((term id/))) ... (((/id)))`

We're trying to come up with a shorthand that expresses a start and end range based on an ID that is natural to type and fits well with existing conventions in AsciiDoc or popular programming languages.

**Comment by apnadkarni** [comment]
Index terms as section attributes works for me. For the shorthand syntax (not sure it's *that* much shorter :) I prefer the heredoc style to the other alternatives.

**Comment by mojavelinux** [comment]
Just to document your option (and a variant of it):

* `(((term >>id))) ... (((id>>)))`
* `(((term <<id))) ... (((id>>)))`

I still prefer the arrows pointing inwards, towards the content.

p.s. It's true the shorthand syntax is not *that* much shorter. Perhaps more accurate to say it's more symbol-oriented ;)

**Comment by apnadkarni** [comment]
Isn't the use of >> potential for confusion with >> being used for the "see other" index reference?

**Comment by mojavelinux** [comment]
Nope, because of the placement at the very end or beginning of the enclosure. As to whether that is confusing for the human, that's a different matter.

We could perhaps use:

* `(((term >id>))) ... (((<id<)))`

**Comment by mojavelinux** [comment]
Another option is:

* `(((term id~>))) ... (((<~id)))`

**Comment by apnadkarni** [comment]
Having ruminated on it, I think your first suggestions works best. 

    (((term >>))) ... stuff (((<< term))

**Comment by mojavelinux** [comment]
Almost. I think we should use an ID instead of repeating the term. Or do you think otherwise?

* `(((term term_id>>))) ... stuff (((<<term_id))`

**Comment by apnadkarni** [comment]
Sorry, what you wrote is what I meant.

**Comment by meisterluk** [comment]
Just a remark: This above-mentioned range syntax is also used by O'Reilly. See [O'Reilly's AsciiDoc Indexing Guidelines](http://chimera.labs.oreilly.com/books/1234000001578/ch02.html).

**Comment by seinecle** [comment]
Another way to consider this issue. So I have this kind of rendering:

![screen](https://user-images.githubusercontent.com/1244100/39912479-25935ea0-54ff-11e8-8e3c-590d6e0787cf.png)

I'd like to have page range 32-38 instead of 32, 33, 34, 35, 36, 37, 38

In the case of a printed book output (`:media: print` or `:media: prepress` set in the doc, see https://github.com/asciidoctor/asciidoctor-pdf/issues/904), where we don't need a link to the entry, then it might be possible in the index generation process to **detect consecutive pages like above and replace them with the corresponding page range**. Without additional attributes in the original asciidoc. What do you think?

**Comment by mojavelinux** [comment]
> detect consecutive pages like above and replace them with the corresponding page range

That's a good idea, though I think it's separate from this one. Can you post that issue in Asciidoctor PDF?
